openzl-crypto/src/merkle_tree/path.rs

```diff
@@ -0,0 +1,1155 @@
+    {
+        self.is_current_with(&Default::default())
+    }
+
+    /// Checks if `self` could represent the [`CurrentInnerPath`] of some tree, using `default`
+    /// as the sentinel value.
+    #[inline]
+    pub fn is_current_with(&self, default: &InnerDigest<C>) -> bool
+    where
+        InnerDigest<C>: PartialEq,
+    {
+        InnerNodeIter::from_leaf::<C>(self.leaf_index)
+            .zip(self.path.iter())
+            .all(move |(node, d)| match node.parity() {
+                Parity::Left => d == default,
+                Parity::Right => true,
+            })
+    }
+
+    /// Computes the root of the merkle tree relative to `base` using `parameters`.
+    #[inline]
+    pub fn root_from_base(&self, parameters: &Parameters<C>, base: InnerDigest<C>) -> Root<C> {
+        Self::fold(parameters, self.leaf_index, base, &self.path)
+    }
+
+    /// Computes the root of the merkle tree relative to `leaf_digest` and its `sibling_digest`
+    /// using `parameters`.
+    #[inline]
+    pub fn root(
+        &self,
+        parameters: &Parameters<C>,
+        leaf_digest: &LeafDigest<C>,
+        sibling_digest: &LeafDigest<C>,
+    ) -> Root<C> {
+        self.root_from_base(
+            parameters,
+            self.leaf_index
+                .join_leaves(parameters, leaf_digest, sibling_digest),
+        )
+    }
+
+    /// Returns `true` if `self` is a witness to the fact that `leaf_digest` is stored in a
+    /// merkle tree with the given `root` and `sibling_digest`.
+    #[inline]
+    pub fn verify_digest(
+        &self,
+        parameters: &Parameters<C>,
+        root: &Root<C>,
+        leaf_digest: &LeafDigest<C>,
+        sibling_digest: &LeafDigest<C>,
+    ) -> bool
+    where
+        InnerDigest<C>: PartialEq,
+    {
+        root == &self.root(parameters, leaf_digest, sibling_digest)
+    }
+
+    /// Returns the folding algorithm for a path with `index` as its starting index.
+    #[inline]
+    fn fold_fn<'d>(
+        parameters: &'d Parameters<C>,
+        mut index: Node,
+    ) -> impl 'd + FnMut(InnerDigest<C>, &'d InnerDigest<C>) -> InnerDigest<C> {
+        move |acc, d| index.into_parent().join(parameters, &acc, d)
+    }
+
+    /// Folds `iter` into a root using the path folding algorithm for [`InnerPath`].
+    #[inline]
+    pub(crate) fn fold<'i, I>(
+        parameters: &'i Parameters<C>,
+        index: Node,
+        base: InnerDigest<C>,
+        iter: I,
+    ) -> InnerDigest<C>
+    where
+        InnerDigest<C>: 'i,
+        I: IntoIterator<Item = &'i InnerDigest<C>>,
+    {
+        iter.into_iter()
+            .fold(base, Self::fold_fn(parameters, index))
+    }
+}
+
+impl<C> Default for InnerPath<C>
+where
+    C: Configuration + ?Sized,
+    InnerDigest<C>: Default,
+{
+    #[inline]
+    fn default() -> Self {
+        let path_length = path_length::<C, _>();
+        let mut path = Vec::with_capacity(path_length);
+        path.resize_with(path_length, InnerDigest::<C>::default);
```

Comment:
How does this differ from `let path = vec![ InnerDigest::<C>::default() ; path_length ]` ?

Reply:
Technically the one in the code is more efficient since it doesn't clone values.